### PR TITLE
Base MDX type tests on real world examples

### PR DESCRIPTION
**Pull request by remcohaszing**
This bases the MDX type tests on real world examples. Different frameworks use different type definitions for the JSX namespace. This is what I envision the tests could look like.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://discord.com/channels/508357248330760243/757992230077333655/1215735407824732170
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

cc @jakebailey

### CODE REVIEW

**Comment by jakebailey** [comment]
ts-eslint's support for project references is not so good; you'd need to set this up just like I did #68908 and put tsconfigs without references at the root.

File: types/mdx/package.json
Comment (jakebailey) [inline types/mdx/package.json:1]: This file should have:

```
"tsconfigs": ["tsconfig.preact.json", "tsconfig.react.json"]
```

And maybe another framework-less tsconfig if desired?

Comment (remcohaszing) [reply types/mdx/package.json:1]: I restored the framework-less tests. It could also be interesting to add additional tests for Vue and Solid.

Are the tsconfigs from the `tsconfigs` array allowed to be in subdirectories? That would solve the whole problem of the editing experience picking up the wrong tsconfig file.

File: types/mdx/package.json
Comment (jakebailey) [inline types/mdx/package.json:10]: `tsconfig.json` isn't the thing to have in here; basically you need three separate configs that only cover the files you want, then the root tsconfig loads _all_ files (including those in the other tsconfigs) which ts-eslint uses to lint and dtslint uses to list the files.

It's a little wonky but changing that will take a bit more restructuring in dtslint and the way we use eslint.

File: types/mdx/package.json
Comment (jakebailey) [reply types/mdx/package.json:1]: They are, but that's accidental. The way the infra works requires that the root tsconfig list all files to lint, so the root one needs to include all files and all other files are totally ignored.

Comment (jakebailey) [reply types/mdx/package.json:1]: No, they're not allowed, oops; I forgot that the restriction is that the tsconfigs be named `tsconfig.*.json` so those child tsconfigs won't be matched.

Comment (jakebailey) [reply types/mdx/package.json:1]: That being said, maybe having them in child dirs that ts-eslint can easily load may improve this... But dtslint still has to change to linting all files in a dir and not just those in the root tsconfig. (That's more challenging.)

File: types/mdx/package.json
Comment (jakebailey) [inline types/mdx/package.json:10]:
```suggestion
    "tsconfigs": ["tsconfig.frameworkless.json", "tsconfig.preact.json", "tsconfig.react.json"],
```

Based on the files in the PR?

File: types/mdx/package.json
Comment (remcohaszing) [reply types/mdx/package.json:1]: I have my doubts about the usefulness of the frameworkless config if some frameworks are actually tested.

File: types/mdx/preact-tests.tsx
Comment (remcohaszing) [inline types/mdx/preact-tests.tsx:21]: I was a bit sad to find out this is necessary. I initially thought this would work, but it doesn’t:

```ts
declare module 'mdx/types' {
  import { JSX } from 'preact/jsx-runtime'
}
```

File: types/mdx/package.json
Comment (jakebailey) [reply types/mdx/package.json:1]: Oh, certainly, I wasn't sure of the value, as I don't really know much about MDX 😄

**Comment by jakebailey** [comment]
I think to fix the current errors, you should go back to `types: []` and instead use reference directives to load react/preact.

File: types/mdx/package.json
Comment (remcohaszing) [reply types/mdx/package.json:1]: What’s mostly relevant here is that it depends on the `JSX` namespace. Each framework has its own implementation of this `JSX` namespace. Lately there have been some developments in TypeScript regarding the `JSX` namespace, and also some frameworks have removed the global `JSX` namspace.

So it’s nice to assert how all this affects the MDX types in different contexts.

**Comment by jakebailey** [comment]
Seems like the test only fails due to the JSX namespace needing to be declared?

**Comment by remcohaszing** [comment]
The JSX namespace is declared in `types/mdx/preact-tests.tsx` via module augmentation. I don’t see why it’s currently not working. I’m pretty sure this worked before.

**Comment by jakebailey** [comment]
The one declared in the preact file is on the preact module, but `index.d.ts` is looking for the one in the globals.

**Comment by jakebailey** [comment]
Seems like it's passing; are you wanting this to get merged in? (It's currently a draft but seems good to me.)

**Comment by remcohaszing** [comment]
I think it’s good, but seeing Preact integration depends on declaration merging, I would like to add some docs for this. Also now that this works, I want to test Vue and Solid integration.

I’m marking it _Ready for review_ to ping other maintainers, see if they also think this is a good idea.

**Comment by weswigham** [comment]
I'm pretty sure if you set `jsxImportSource` to `preact`, the preact example won't need anything special, type-wise, at least for `"jsx": "react-jsx"`

**Comment by remcohaszing** [comment]
> I'm pretty sure if you set `jsxImportSource` to `preact`, the preact example won't need anything special, type-wise, at least for `"jsx": "react-jsx"`

This used to be the case, but with frameworks moving away from the global `JSX` namespace, it no longer is.

Also DefinitelyTypes doesn’t allow specifying `jsxImportSource`. The classic runtime works for me so far, but maybe it should be allowed.

**Comment by sheetalkamat** [comment]
@jakebailey @sandersn @weswigham is this good to go or does this need some work as suggested by @weswigham in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68955#issuecomment-1997906793 to support supplying `jsxImportSource` or similar ?

**Comment by jakebailey** [comment]
This PR is fine from my perspective, I believe we were just waiting for feedback from other MDX maintainers.

**Review by sheetalkamat — DISMISSED** [review]
Marking this as  change needed so that its not on maintainer board. please merge main when one of MDX maintainer approves so we can approve this.

**Comment by jakebailey** [comment]
FWIW you can just move this over to the "needs review" column and it'll "bless" the PR to say "maintainer said okay", I'm pretty sure.

File: types/mdx/preact-tests.tsx
Comment (ChristianMurphy) [inline types/mdx/preact-tests.tsx:339]: is there something about input that makes it a better test than `video`?

**Review by ChristianMurphy — APPROVED** [review]
I don't have access to the discord that initiated this PR to have the context.
Reading between the lines a bit, it seems to focus on adding `preact` testing?

I'm all for supporting the types of other frameworks.
But don't use them enough to offer much constructive review on this PR.
I'll approved to unblock if others are comfortable.

/cc @wooorm who have more thoughts on this

**Comment by jakebailey** [comment]
> I don't have access to the discord that initiated this PR to have the context.
> Reading between the lines a bit, it seems to focus on adding `preact` testing?

The context is that I added to DT the ability to check with multiple different tsconfigs, which means projects can check with different settings and references. This was mainly for node to check with/without DOM types, but can also be used for MDX to test against react, preact, etc.

**Comment by ChristianMurphy** [comment]
Thanks for the context @jakebailey! That helps

**Review by jakebailey — APPROVED** [review]
Doesn't seem like there's going to be other feedback; going to approve this and you can feel free to merge if you believe that it's all good.

File: types/mdx/preact-tests.tsx
Comment (remcohaszing) [reply types/mdx/preact-tests.tsx:339]: Yes, apparently input props are assignable to image props in Preact.

**Comment by remcohaszing** [comment]
It looks like both Solid and Vue only work with the automatic runtime types, which are not allowed by dtslint. I’ll merge as-is.

**Comment by remcohaszing** [comment]
Ready to merge

**Pull request merged**
